rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,113 @@
+---
+feature: nix-language-version
+start-date: 2022-12-12
+author: @fricklerhandwerk
+co-authors: @thufschmitt @Ericson2314
+shepherd-team: 
+shepherd-leader: 
+related-issues: https://github.com/NixOS/nix/issues/7255
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.
+
+# Motivation
+[motivation]: #motivation
+
+Currently it's impossible to introduce backwards-incompatible changes to the Nix language without breaking existing setups.
+This proposal is first step towards overcoming that limitation.
+
+# Detailed design
+[design]: #detailed-design
+
+Introduce a magic comment in the first line of a Nix file:
+
+    #? Nix <version>
+
+where `<version>` is a released version of Nix the given file is intended to work with.
```

Comment:
Should it be an exact version like `2.12.0` or should bounds like `>= 2.12.0` or `2.12.*` also be allowed?

Or is the intention for this to be an arbitrary text string? If so this sounds a lot like the [User Agent](https://en.wikipedia.org/wiki/User_agent) war, I think it should definitely be a very well-defined field here. I think basing this on a single increasing number like `builtins.langVersion` would be best, this also paves the way towards a Nix language standard.

Reply:
Could you elaborate on what you mean by User Agent war? How does it relate to providing information about the grammar used in a given file?

In any case, this RFC explicitly states twice that semantics are out of scope. The question purely if we want to encode that information or not, and only superficially how to do that. Everything else is left to define in future iterations.

Reply:
I imagine that it should be an exact version. I'm thinking of Rust "editions" as a model. If you specify `Nix 2.12` any new breaking feature are not enabled when parsing the module. When you feel like it you update to the new syntax and update your tag line to `Nix 2.14`. 

That being said I think it probably makes more sense to learn from the Rust approach, it has worked out really well. Maybe instead of Nix versions we should just occasionally cut a new "edition" with a batch of breaking changes. Of course old editions are still supported in all new Nix builds (at least for a very long time).

Reply:
Regarding User Agent: [wiki](https://en.wikipedia.org/wiki/User_agent#User_agent_spoofing) or [Chrome phasing out User Agent support](https://news.ycombinator.com/item?id=22685018) or some [history of User Agent](https://webaim.org/blog/user-agent-string-history/)

> In any case, this RFC explicitly states twice that semantics are out of scope.

Then I guess I don't agree with that. I don't think it make sense to introduce new syntax without defining the semantics of it.

Reply:
Alright, I concede I'm just reluctant to let the RFC scope spiral out of control. I'll add all semantics proposals, but please be gentle with me.